sr_coex: Add support for Wi-Fi and BLE co-existence (#9778)
nrfconnect/sdk-nrf: samples/wifi/sr_coex/README.rst

***************
Program the nRF5340 DK:

* Open a new terminal in test PC. Navigate to ``<ncs code>/nrf/samples/bluetooth/throughput/``

.. code-block:: console

   $ west flash --dev-id <device-id> --hex-file build/zephyr/merged_domains.hex

Program the nRF7002 DK:

* Open a new terminal in test PC. Navigate to ``<ncs code>/nrf/samples/wifi/sr_coex/``

.. code-block:: console

   $ west flash --dev-id <device-id> --hex-file build/zephyr/merged_domains.hex

Testing
*******

Running coexistence sample test cases require additional software such as Wi-Fi iperf application. When coexistence sample runs Wi-Fi UDP throughput in client mode, peer device runs UDP throughput in server mode using command

.. code-block:: console

   $ iperf -s -i 1 -u

Use iperf version 2.0.5. For more details Refer `Network Traffic Generator`_.

+---------------+--------------+--------------------------------------------------------+
| Test case     | Coexistence  | Test procedure                                         |
|               |              |                                                        |
+===============+==============+========================================================+

> | Wi-Fi only    | N.A          | Run Wi-Fi iperf in server mode on test PC.             |
> | throughput    |              | Program coexistence sample application on nRF7002 DK.  |
> +---------------+--------------+--------------------------------------------------------+
> | BLE only      | N.A          | Program BLE throughput application on nRF5340 DK and   |
> | throughput    |              | Select role as peripheral.                             |
> |               |              | Program coexistence sample application on nRF7002 DK.  |
> +---------------+--------------+--------------------------------------------------------+
> | Wi-Fi and BLE | Disabled/    | Run Wi-Fi iperf in server mode on test PC.             |
> | combined      | Enabled      | Program BLE throughput application on nRF5340 DK and   |
> | throughput    |              | Select role as peripheral.                             |
> |               |              | Program coexistence sample application on nRF7002 DK.  |

Reviewer: ```suggestion
| Wi-Fi only    | NA           | Run Wi-Fi **iperf** in server mode on test PC.         |
| throughput    |              | Program coexistence sample application on nRF7002 DK.  |
+---------------+--------------+--------------------------------------------------------+
| Bluetooth LE only      | NA          | Program BLE throughput application on nRF5340 DK and   |
| throughput    |              | select role as peripheral.                             |
|               |              | Program coexistence sample application on nRF7002 DK.  |
+---------------+--------------+--------------------------------------------------------+
| Wi-Fi and BLE | Disabled/    | Run Wi-Fi **iperf** in server mode on test PC.             |
| combined      | Enabled      | Program Bluetooth LE throughput application on nRF5340 DK and   |
| throughput    |              | select role as peripheral.                             |
|               |              | Program coexistence sample application on nRF7002 DK.  |
```
Again, messed up the table, I'm afraid. Just focused on the text. Bear with me.

+---------------+--------------+--------------------------------------------------------+

> Wi-Fi throughput result appears on test PC terminal on which iperf server is run. BLE throughput result appears on minicom terminal connected to nRF5340 DK.

Reviewer: ```suggestion
The Wi-Fi throughput result appears on the test PC terminal on which **iperf** server is run.
The Bluetooth LE throughput result appears on the minicom terminal connected to the nRF5340 DK.
```

Results
*******

> Summary of results obtained when coexistence tests are run for different Wi-Fi operating bands and antenna configurations with Wi-Fi mode set to 802.11n are given below. These tests are run keeping test setup inside an RF shield box. So, these results are representative and might change with changes in test setup.

Reviewer: ```suggestion
The following table collects a summary of results obtained when coexistence tests are run for different Wi-Fi operating bands and antenna configurations with Wi-Fi mode set to 802.11n.
The tests are run with the test setup inside an RF shield box.
Therefore, the results are representative and might change with adjustments in the test setup.
```

> **Wi-Fi in 2.4GHz**

Reviewer: ```suggestion
Wi-Fi in 2.4 GHz
================
```

> *Separate antennas*

Reviewer: ```suggestion
Separate antennas
```

> 

Reviewer: BLE is a banned abbreviation. Change it to Bluetooth LE in all tables.

+----------------------+--------------------+-----------------+
| Test case            | Wi-Fi UDP Tx       | BLE throughput  |
|                      | throughput in Mbps | in kbps         |
+======================+====================+=================+
| Wi-Fi only,          | 10.2               | N.A             |
| client (UDP Tx)      |                    |                 |
+----------------------+--------------------+-----------------+
| BLE only,            | N.A                | 1107            |
| central              |                    |                 |
+----------------------+--------------------+-----------------+
| Wi-Fi and BLE,       | 9.9                | 145             |
| coexistence disabled |                    |                 |
+----------------------+--------------------+-----------------+
| Wi-Fi and BLE,       | 8.3                | 478             |
| coexistence enabled  |                    |                 |
+----------------------+--------------------+-----------------+

> *Shared antenna*

Reviewer: ```suggestion
Shared antenna
```

+----------------------+--------------------+-----------------+
| Test case            | Wi-Fi UDP Tx       | BLE throughput  |
|                      | throughput in Mbps | in kbps         |
+======================+====================+=================+
| Wi-Fi only,          | 10.2               | N.A             |
| client (UDP Tx)      |                    |                 |
+----------------------+--------------------+-----------------+
| BLE only,            | N.A                | 1219            |
| central              |                    |                 |
+----------------------+--------------------+-----------------+
| Wi-Fi and BLE,       | 10.2               | 29              |
| coexistence disabled |                    |                 |
+----------------------+--------------------+-----------------+
| Wi-Fi and BLE,       | 6.2                | 749             |
| coexistence enabled  |                    |                 |
+----------------------+--------------------+-----------------+

> **Wi-Fi in 5GHz**

Reviewer: ```suggestion
Wi-Fi in 5 GHz
=============
```

> *Separate antennas*

Reviewer: ```suggestion
Separate antennas
```

+----------------------+--------------------+-----------------+
| Test case            | Wi-Fi UDP Tx       | BLE throughput  |
|                      | throughput in Mbps | in kbps         |
+======================+====================+=================+
| Wi-Fi only,          | 10.2               | N.A             |
| client (UDP Tx)      |                    |                 |
+----------------------+--------------------+-----------------+
| BLE only,            | N.A                | 1139            |
| central              |                    |                 |
+----------------------+--------------------+-----------------+
| Wi-Fi and BLE,       | 10.2               | 1188            |
| coexistence disabled |                    |                 |
+----------------------+--------------------+-----------------+
| Wi-Fi and BLE,       | 10.2               | 1208            |
| coexistence enabled  |                    |                 |
+----------------------+--------------------+-----------------+

> *Shared antenna*

Reviewer: ```suggestion
Shared antenna
```

+----------------------+--------------------+-----------------+
| Test case            | Wi-Fi UDP Tx       | BLE throughput  |
|                      | throughput in Mbps | in kbps         |
+======================+====================+=================+
| Wi-Fi only,          | 10.2               | N.A             |
| client (UDP Tx)      |                    |                 |
+----------------------+--------------------+-----------------+
| BLE only,            | N.A                | 1180            |
| central              |                    |                 |
+----------------------+--------------------+-----------------+
| Wi-Fi and BLE,       | 10.2               | 1177            |
| coexistence disabled |                    |                 |
+----------------------+--------------------+-----------------+
| Wi-Fi and BLE,       | 10.2               | 1191            |
| coexistence enabled  |                    |                 |
+----------------------+--------------------+-----------------+

Sample output
*************

> Screenshots of coexistence test results obtained for separate antenna configuration with Wi-Fi mode set to 802.11n are given below. These tests are run when WLAN connected to an AP in 2.4GHz band. In these, left-side result shows Wi-Fi throughput that appears on test PC terminal in which Wi-Fi iperf server is run and right-side result shows BLE throughput that appears on minicom terminal in which BLE throughput sample is run.

Reviewer: ```suggestion
The following screenshots show coexistence test results obtained for separate antenna configuration with Wi-Fi mode set to 802.11n.
These tests were run with WLAN connected to an AP in 2.4 GHz band.
In the images, left-side result shows Wi-Fi throughput that appears on a test PC terminal in which Wi-Fi **iperf** server is run and right-side result shows Bluetooth LE throughput that appears on a minicom terminal in which the Bluetooth LE throughput sample is run.
```

.. figure:: images/wlan.jpg
     :width: 711px
     :align: center
     :alt: Wi-Fi only throughput

>      Wi-Fi only throughput 10.2Mbps

Reviewer: ```suggestion
     Wi-Fi only throughput 10.2 Mbps
```

.. figure:: images/ble.jpg
     :width: 711px
     :align: center
     :alt: BLE only throughput

>      BLE only throughput: 1107kbps

Reviewer: ```suggestion
     Bluetooth LE only throughput: 1107 kbps
```

.. figure:: images/wlan_ble_cd.jpg
     :width: 711px
     :align: center
     :alt: Wi-Fi and BLE CD

>      Wi-Fi and BLE throughput, coexistence disabled: Wi-Fi 9.9Mbps and BLE 145kbps

Reviewer: ```suggestion
     Wi-Fi and Bluetooth LE throughput, coexistence disabled: Wi-Fi 9.9 Mbps and BLE 145 kbps
```

.. figure:: images/wlan_ble_cd.jpg
     :width: 711px
     :align: center
     :alt: Wi-Fi and BLE CE

>      Wi-Fi and BLE throughput, coexistence enabled: Wi-Fi 8.3Mbps and BLE 478kbps

Reviewer: ```suggestion
     Wi-Fi and Bluetooth LE throughput, coexistence enabled: Wi-Fi 8.3 Mbps and Bluetooth LE 478 kbps
```

> 

Reviewer: Remove extra empty lines. One is enough.

Dependencies
************

This sample uses the following `sdk-nrfxlib`_ library:

* :ref:`nrfxlib:nrf_security`

.. _Threads: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/kernel/services/threads/index.html#threads

> .. _Network Traffic Generator: Network Traffic Generator — Zephyr Project Documentation

Reviewer: Instead of introducing the link targets here, insert them in doc/nrf/links.txt

Test and evaluation setup
*************************

The following figure shows a reference test setup and connections between the following devices:

* :term:`Device Under Test (DUT)` (nRF7002 DK on which the coexistence sample runs)
* Wi-Fi peer device (access point with test PC that runs **iperf**)
* Bluetooth LE peer device (nRF5340 DK on which Bluetooth LE throughput sample runs)

.. figure:: images/wifi_coex.svg
     :width: 711px
     :align: center
     :alt: Wi-Fi Coex test setup

     Wi-Fi coexistence reference test and evaluation setup

The following table provides more details on the sample or application that runs on DUT and peer devices.
This setup is kept in a shielded test enclosure box (for example a Ramsey box).

+--------------+----------------+------------------------------------------------------------------------------+
| Device       | Application    |                             Details                                          |
+==============+================+==============================================================================+

> | nRF7002 DK   | coex sample    | The sample runs Wi-Fi throughputs, Bluetooth LE throughputs                  |

Reviewer 2: Bluetooth LE coexistence sample

* Bluetooth LE throughput only
* Concurrent Wi-Fi and Bluetooth LE throughput (with coexistence enabled and disabled mode)

To enable different test modes, set up the following configuration parameters in the :file:`prj.conf` file:

* Antenna configuration: Use the :kconfig:option:`CONFIG_COEX_SEP_ANTENNAS` Kconfig option to select the antenna configuration.
  Set it to ``y`` to enable separate antennas and ``n`` to enable shared antenna.
* Test modes: Use the following Kconfig options to select the required test case:

  * :kconfig:option:`CONFIG_TEST_TYPE_WLAN_ONLY` for Wi-Fi only test
  * :kconfig:option:`CONFIG_TEST_TYPE_BLE_ONLY` for Bluetooth LE only test
  * :kconfig:option:`CONFIG_TEST_TYPE_WLAN_BLE` for concurrent Wi-Fi and Bluetooth LE test.

  Based on the required test, set only one of these to ``y`` while the other two have to be set to ``n``.
* Test duration: Use the :kconfig:option:`CONFIG_WIFI_TEST_DURATION` Kconfig option to set the duration of the Wi-Fi test and :kconfig:option:`CONFIG_BLE_TEST_DURATION` for the Bluetooth LE test.
  The units are in milliseconds.
  For example, to set the tests for 20 seconds, set the respective values to ``20000``.

>   For the concurrent Wi-Fi and Bluetooth LE test, make sure that both of these are set to the same duration to ensure maximum overlap.

Reviewer 2: ```suggestion
  For the concurrent Wi-Fi and Bluetooth LE test, make sure that both are set to the same duration to ensure maximum overlap.
```

* Build with coexistence enabled:

.. code-block:: console

   west build -b nrf7002dk_nrf5340_cpuapp -- -DCONFIG_MPSL_CX=y -Dhci_rpmsg_CONFIG_MPSL_CX=y

The generated hex file to be used is :file:`build/zephyr/merged_domains.hex`.

Use the Bluetooth throughput sample from the :file:`nrf/samples/bluetooth/throughput` folder on the peer nRF5340 DK device.

* Build for the nRF5340 DK:

.. code-block:: console

   west build -b nrf5340dk_nrf5340_cpuapp

Set up the following configuration parameters in the :file:`prj_nrf5340dk_nrf5340_cpuapp.conf` file:

* File or time-based throughput: Use ``BT_THROUGHPUT_FILE`` to select file or time-based throughput test.
  Set it to ``n`` to enable time-based throughput test only when running Bluetooth LE throughput in central role.
* Test duration: Use ``BT_THROUGHPUT_DURATION`` to set the duration of the Bluetooth LE throughput test only when running Bluetooth LE throughput in central role.
  The units are in milliseconds.

.. note::
   Use the same test duration value for :kconfig:option:`CONFIG_WIFI_TEST_DURATION`, :kconfig:option:`CONFIG_BLE_TEST_DURATION` and ``BT_THROUGHPUT_DURATION``.

Connecting to and programming DKs

> =================================

Reviewer 2: Could be removed as the subheadings are there.

> Connecting to DKs
> *****************

Reviewer 2: ```suggestion
Connecting to DKs
=================
```

> After the DKs are connected to the test PC through USB connectors and powered on, open a suitable terminal and run the following command:

Reviewer 2: ```suggestion
After the DKs are connected to the test PC through USB connectors and powered on, open a suitable terminal, and run the following command:
```

.. code-block:: console

   $ nrfjprog --com
   1050043161         /dev/ttyACM0    VCOM0
   1050043161         /dev/ttyACM1    VCOM1
   1050724225         /dev/ttyACM2    VCOM0
   1050724225         /dev/ttyACM3    VCOM1
   $

In this example, ``1050043161`` is the device ID of the nRF5340 DK and ``1050724225`` is device ID of the nRF7002 DK.

While connecting to a particular board, use the ttyACMx corresponding to VCOMx.
In the example, use ttyACM1 to connect to the board with device ID ``1050043161``.
Similarly, use ttyACM3 connect to the board with device ID ``1050724225``.

.. code-block:: console

   $ minicom -D /dev/ttyACM1 -b 115200
   $ minicom -D /dev/ttyACM3 -b 115200

> Programming DKs
> ***************

Reviewer 2: ```suggestion
Programming DKs
===============
```

| Wi-Fi and     | Disabled/    | Run Wi-Fi **iperf** in server mode on test PC.         |
| Bluetooth LE  | Enabled      | Program Bluetooth LE throughput application on nRF5340 |
| combined      |              | DK and select role as peripheral.                      |
| throughput    |              | Program coexistence sample application on nRF7002 DK.  |
+---------------+--------------+--------------------------------------------------------+

The Wi-Fi throughput result appears on the test PC terminal on which **iperf** server is run.
The Bluetooth LE throughput result appears on the minicom terminal connected to the nRF5340 DK.

Results
*******

The following table collects a summary of results obtained when coexistence tests are run for different Wi-Fi operating bands and antenna configurations with Wi-Fi mode set to 802.11n.
The tests are run with the test setup inside an RF shield box.
Therefore, the results are representative and might change with adjustments in the test setup.

Wi-Fi in 2.4 GHz
================

Separate antennas

+------------------------+--------------------+-------------------+
| Test case              | Wi-Fi UDP Tx       | Bluetooth LE      |

> |                        | throughput in Mbps | throughputin kbps |

Reviewer 2: ```suggestion
|                        | throughput in Mbps | throughput in kbps |
```

+========================+====================+===================+
| Wi-Fi only,            | 10.2               | N.A               |
| client (UDP Tx)        |                    |                   |
+------------------------+--------------------+-------------------+
| Bluetooth LE only,     | N.A                | 1107              |
| central                |                    |                   |
+------------------------+--------------------+-------------------+
| Wi-Fi and Bluetooth LE,| 9.9                | 145               |
| coexistence disabled   |                    |                   |
+------------------------+--------------------+-------------------+
| Wi-Fi and Bluetooth LE,| 8.3                | 478               |
| coexistence enabled    |                    |                   |
+------------------------+--------------------+-------------------+

Shared antenna

+------------------------+--------------------+-------------------+
| Test case              | Wi-Fi UDP Tx       | Bluetooth LE      |

> |                        | throughput in Mbps | throughputin kbps |

Reviewer 2: ```suggestion
|                        | throughput in Mbps | throughput in kbps |
```

+========================+====================+===================+
| Wi-Fi only,            | 10.2               | N.A               |
| client (UDP Tx)        |                    |                   |
+------------------------+--------------------+-------------------+
| Bluetooth LE only,     | N.A                | 1219              |
| central                |                    |                   |
+------------------------+--------------------+-------------------+
| Wi-Fi and Bluetooth LE,| 10.2               | 29                |
| coexistence disabled   |                    |                   |
+------------------------+--------------------+-------------------+
| Wi-Fi and Bluetooth LE,| 6.2                | 749               |
| coexistence enabled    |                    |                   |
+------------------------+--------------------+-------------------+

Wi-Fi in 5 GHz
=============

Separate antennas

+------------------------+--------------------+-------------------+
| Test case              | Wi-Fi UDP Tx       | Bluetooth LE      |

> |                        | throughput in Mbps | throughputin kbps |

Reviewer 2: ```suggestion
|                        | throughput in Mbps | throughput in kbps |
```

+========================+====================+===================+
| Wi-Fi only,            | 10.2               | N.A               |
| client (UDP Tx)        |                    |                   |
+------------------------+--------------------+-------------------+
| Bluetooth LE only,     | N.A                | 1139              |
| central                |                    |                   |
+------------------------+--------------------+-------------------+
| Wi-Fi and Bluetooth LE,| 10.2               | 1188              |
| coexistence disabled   |                    |                   |
+------------------------+--------------------+-------------------+
| Wi-Fi and Bluetooth LE,| 10.2               | 1208              |
| coexistence enabled    |                    |                   |
+------------------------+--------------------+-------------------+

Shared antenna

+------------------------+--------------------+-------------------+
| Test case              | Wi-Fi UDP Tx       | Bluetooth LE      |

> |                        | throughput in Mbps | throughputin kbps |

Reviewer 2: ```suggestion
|                        | throughput in Mbps | throughput in kbps |
```

The sample uses standard Zephyr threads.
The threads are configured with non-negative priority (pre-emptible thread).
For details on threads and scheduling, refer to `Threads`_.

Building and running
********************

.. |sample path| replace:: :file:`samples/wifi/sr_coex`

.. include:: /includes/build_and_run_ns.txt

The sample can be built for the following configurations:

* Wi-Fi throughput only
* Bluetooth LE throughput only
* Concurrent Wi-Fi and Bluetooth LE throughput (with coexistence enabled and disabled mode)

To enable different test modes, set up the following configuration parameters in the :file:`prj.conf` file:

* Antenna configuration: Use the :kconfig:option:`CONFIG_COEX_SEP_ANTENNAS` Kconfig option to select the antenna configuration.
  Set it to ``y`` to enable separate antennas and ``n`` to enable shared antenna.
* Test modes: Use the following Kconfig options to select the required test case:

  * :kconfig:option:`CONFIG_TEST_TYPE_WLAN_ONLY` for Wi-Fi only test
  * :kconfig:option:`CONFIG_TEST_TYPE_BLE_ONLY` for Bluetooth LE only test

>   * :kconfig:option:`CONFIG_TEST_TYPE_WLAN_BLE` for concurrent Wi-Fi and Bluetooth LE test.
> 
>   Based on the required test, set only one of these to ``y`.

Reviewer 2: ```suggestion
  * :kconfig:option:`CONFIG_TEST_TYPE_WLAN_BLE` for concurrent Wi-Fi and Bluetooth LE test.

    Based on the required test, set only one of these to ``y`.
```

  The units are 1.25 milliseconds.
  For example, ``CONFIG_INTERVAL_MIN=80`` corresponds to an interval of 100 ms (80 x 1.25).
* Wi-Fi connection: Set the following options appropriately as per the credentials of the access point used for this testing:

  * :kconfig:option:`CONFIG_STA_SAMPLE_SSID`
  * :kconfig:option:`CONFIG_STA_SAMPLE_PASSWORD`
  * ``STA_KEY_MGMT_*``
  * :kconfig:option:`CONFIG_NET_CONFIG_PEER_IPV4_ADDR`

.. note::
   You can also use ``menuconfig`` to enable the ``Key management`` option.

See :ref:`zephyr:menuconfig` in the Zephyr documentation for instructions on how to run ``menuconfig``.

To build for the nRF7002 DK, use the ``nrf7002dk_nrf5340_cpuapp`` build target.
The following are examples of the CLI commands:

* Build with coexistence disabled:

.. code-block:: console

   west build -b nrf7002dk_nrf5340_cpuapp -- -DCONFIG_MPSL_CX=n -Dhci_rpmsg_CONFIG_MPSL_CX=n

* Build with coexistence enabled:

.. code-block:: console

   west build -b nrf7002dk_nrf5340_cpuapp -- -DCONFIG_MPSL_CX=y -Dhci_rpmsg_CONFIG_MPSL_CX=y

> The generated hex file to be used is :file:`build/zephyr/merged_domains.hex`.

Reviewer 2: ```suggestion
The generated HEX file to be used is :file:`build/zephyr/merged_domains.hex`.
```

+------------------------+--------------------+--------------------+

Sample output
=============

The following screenshots show coexistence test results obtained for separate antenna configuration with Wi-Fi mode set to 802.11n.
These tests were run with WLAN connected to an AP in 2.4 GHz band.
In the images, left-side result shows Wi-Fi throughput that appears on a test PC terminal in which Wi-Fi **iperf** server is run and right-side result shows Bluetooth LE throughput that appears on a minicom terminal in which the Bluetooth LE throughput sample is run.

.. figure:: images/wlan.jpg
     :width: 711px
     :align: center
     :alt: Wi-Fi only throughput

     Wi-Fi only throughput 10.2 Mbps

.. figure:: images/ble.jpg
     :width: 711px
     :align: center
     :alt: Bluetooth LE only throughput

     Bluetooth LE only throughput: 1107 kbps

.. figure:: images/wlan_ble_cd.jpg
     :width: 711px
     :align: center
     :alt: Wi-Fi and Bluetooth LE CD

     Wi-Fi and Bluetooth LE throughput, coexistence disabled: Wi-Fi 9.9 Mbps and Bluetooth LE 145 kbps

.. figure:: images/wlan_ble_ce.jpg
     :width: 711px
     :align: center
     :alt: Wi-Fi and Bluetooth LE CE

> 

Reviewer 2: @richabp, FYI the images might need rework (as discussed offline) and might need to be moved into the common images folder.
